dipy/align/streamwarp.py

```diff
@@ -0,0 +1,287 @@
+
+    alpha : float, optional
+        Represents the trade-off between regularizing the deformation and
+        having points match very closely. Lower value of alpha means high
+        deformations (default 0.3)
+
+    beta : int, optional
+        Represents the strength of the interaction between points
+        Gaussian kernel size (default 20)
+
+    affine : boolean, optional
+        If False, use rigid registration as starting point (default True)
+
+    Returns
+    -------
+    deformed_bundle : Streamlines
+        Nonlinearly moved bundle (warped bundle)
+
+    moving_aligned : Streamlines
+        Linearly moved bundle (affinely moved)
+
+    dist : np.ndarray
+        Float array containing distance between moving and static bundle
+
+    matched_pairs : np.ndarray
+        Int array containing streamline correspondences between two bundles
+
+    warp : np.ndarray
+        Nonlinear warp map generated by BundleWarp
+
+    References
+    ----------
+    .. [Chandio2023] Chandio et al. "BundleWarp, streamline-based nonlinear
+            registration of white matter tracts." bioRxiv (2023): 2023-01.
+    """
+    if alpha <= 0.01:
+        warnings.warn("Using alpha<=0.01 will result in extreme deformations")
+
+    if average_bundle_length(static) <= 50:
+        beta = 10
+
+    x0 = 'affine' if affine else 'rigid'
+    moving_aligned, _, _, _ = slr_with_qbx(static, moving, x0=x0,
+                                           rm_small_clusters=0)
+
+    if dist is not None:
+        print("using pre-computed distances")
+    else:
+        dist = distance_matrix_mdf(static, moving_aligned).T
+
+    matched_pairs = np.zeros((len(moving), 2))
+    matched_pairs1 = np.asarray(linear_sum_assignment(dist)).T
+
+    for mt in matched_pairs1:
+        matched_pairs[mt[0]] = mt
+
+    num = len(matched_pairs1)
+
+    all_pairs = list(matched_pairs1[:, 0])
+    all_matched = False
+
+    while all_matched is False:
+
+        num = len(all_pairs)
+
+        if num < len(moving):
+
+            ml = find_missing(all_pairs, moving)
+
+            dist2 = dist[:][ml]
+
+            # dist2 has distance among unmatched streamlines of moving bundle
+            # and all static bundle's streamlines
+
+            matched_pairs2 = np.asarray(linear_sum_assignment(dist2)).T
+
+            for i in range(matched_pairs2.shape[0]):
+                matched_pairs2[i][0] = ml[matched_pairs2[i][0]]
+
+            for mt in matched_pairs2:
+                matched_pairs[mt[0]] = mt
+
+            all_pairs.extend(matched_pairs2[:, 0])
+
+            num2 = num + len(matched_pairs2)
+            if num2 == len(moving):
+                all_matched = True
+                num = num2
+        else:
+            all_matched = True
+
+    deformed_bundle = Streamlines([])
+    warp = []
+
+    # Iterate over each pair of streamlines and deform them
+    # Append deformed streamlines in deformed_bundle
+
+    for _, pairs in enumerate(matched_pairs):
```

Comment:
Why not just pairs in matched_pairs?

Reply:
It returns the index and item
I don't need an index

Reply:
exactly. So unless I am missing something
`for pairs in matched_pairs:`
is the same thing right?

Reply:
pairs is a list. Each item is a list of length 2.

Reply:
so you can do:

```python
for pair_1, pair_2 in matched_pairs:
```

Reply:
works fine as it is though 😄